From 3bf28d03a025f7cd7fc174af4102405090dba24c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Timo=20Ter=C3=A4s?= <timo.teras@iki.fi>
Date: Fri, 2 Aug 2024 13:44:35 +0300
Subject: [PATCH 1/4] db: remove unused field data_csum in struct install_ctx

---
 src/database.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/database.c b/src/database.c
index a555fcc..3697116 100644
--- a/src/database.c
+++ b/src/database.c
@@ -64,7 +64,6 @@ struct install_ctx {
 	unsigned int missing_checksum : 1;
 
 	struct apk_db_dir_instance *diri;
-	struct apk_checksum data_csum;
 	struct apk_extract_ctx ectx;
 
 	apk_progress_cb cb;
-- 
2.45.2

