--- a/drivers/net/dsa/qca8k.c
+++ b/drivers/net/dsa/qca8k.c
@@ -1145,6 +1145,8 @@ qca8k_sw_probe(struct mdio_device *mdiodev)
 	struct qca8k_priv *priv;
 	u32 id;
 
+	pr_info("QCA8K INIT");
+
 	/* allocate the private data struct so that we can probe the switches
 	 * ID register
 	 */
@@ -1178,8 +1180,10 @@ qca8k_sw_probe(struct mdio_device *mdiodev)
 	id = qca8k_read(priv, QCA8K_REG_MASK_CTRL);
 	id >>= QCA8K_MASK_CTRL_ID_S;
 	id &= QCA8K_MASK_CTRL_ID_M;
-	if (id != data->id)
-		return -ENODEV;
+	if (id != data->id) {
+		pr_info("QCA8K FOUND ID %d EXPECTED %d", id, data->id);
+		// return -ENODEV;
+	}
 
 	priv->ds = dsa_switch_alloc(&mdiodev->dev, QCA8K_NUM_PORTS);
 	if (!priv->ds)
@@ -1191,6 +1195,8 @@ qca8k_sw_probe(struct mdio_device *mdiodev)
 	mutex_init(&priv->reg_mutex);
 	dev_set_drvdata(&mdiodev->dev, priv);
 
+	pr_info("QCA8K ALL CHECK PASSED");
+
 	return dsa_register_switch(priv->ds);
 }
 
